src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/reader/ReaderMetricsProcessor.java

```diff
@@ -201,8 +219,46 @@ public void trimOldSnapshots() throws Exception {
     trimMap(shardRqMetricsMap, RQ_SNAPSHOTS);
     trimMap(httpRqMetricsMap, HTTP_RQ_SNAPSHOTS);
     trimMap(masterEventMetricsMap, MASTER_EVENT_SNAPSHOTS);
-    trimDatabases(
-        metricsDBMap, MAX_DATABASES, PluginSettings.instance().shouldCleanupMetricsDBFiles());
+
+    boolean deleteDBFiles = PluginSettings.instance().shouldCleanupMetricsDBFiles();
+    // Cleanup all but the 2 most recent metricsDB files from metricsDBMap. The most recent metricsDB files needs to be
+    // retained for future metrics query handling, the second most recent metricsDB file needs to be retained in case
+    // any metrics query handler just got access to it right before the most recent metricsDB file was available.
+    while (metricsDBMap.size() > MAX_DATABASES) {
+      Map.Entry<Long, MetricsDB> oldestEntry = metricsDBMap.pollFirstEntry();
+      if (oldestEntry != null) {
+        Long key = oldestEntry.getKey();
+        MetricsDB value = oldestEntry.getValue();
+        value.remove();
+        if (deleteDBFiles && !batchMetricsDBSet.contains(key)) {
+          value.deleteOnDiskFile();
+        }
+      }
+    }
+    // Flush any tracking batch metrics if batch metrics is disabled. Note, in order to ensure that batch metrics
+    // consumers have had at least one cycle to use any metrics they may be holding, this flush is done before
+    // re-reading the config file to update the state of the batch metrics feature.
+    if (!batchMetricsEnabled && !batchMetricsDBSet.isEmpty()) {
+      if (deleteDBFiles) {
+        for (Long timestamp : batchMetricsDBSet) {
+          if (!metricsDBMap.containsKey(timestamp)) {
+            MetricsDB.deleteOnDiskFile(timestamp);
+          }
+        }
+      }
+      batchMetricsDBSet.clear();
+    }
+    readBatchMetricsEnabledFromConf();
```

Comment:
Although we are playing safe here by not reading the config before, but I don't think we are completely avoiding the race here. Say, if reading all the files in the retention period is taking too long, (multiple iterations of this call), then there is still a chance that the API read thread tries to read a file that has been deleted. So, we should definitely handle that exception, in any case.

Reply:
Right, the metrics processor code gives the query handler's a ~5s grace period before deleting stale metricsdb files (see the javadoc on getBatchMetrics). For the typical case (when batch metrics is not about to be disabled), this will not be an issue since the batch metrics query handler reads data from the older files first. For the case when batch metrics is about to be disabled, even newer files are at risk of deletion. However, this disabling of batch metrics is not expected to be a use case, there is a 100,800 datapoints limit on the query handler, there's not much data to say that this grace period will be an issue given the datapoint limit, and the query handler properly handles missing files.